internal/driver/custommetadata.go

```diff
@@ -0,0 +1,124 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2022 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package driver
+
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/edgexfoundry/go-mod-core-contracts/v2/errors"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/v2/models"
+)
+
+func (onvifClient *OnvifClient) initCustomMetadata(device contract.Device) contract.Device {
+	if _, found := device.Protocols[CustomMetadata]; !found {
+		saveOnvifProtocol := device.Protocols[OnvifProtocol]
+		device.Protocols = map[string]contract.ProtocolProperties{
+			OnvifProtocol:  saveOnvifProtocol,
+			CustomMetadata: {},
+		}
+	}
+	return device
+}
+
+// setCustomMetadata will return a map containing the fields provided in the call to the function
+func (onvifClient *OnvifClient) setCustomMetadata(device contract.Device, data []byte) (contract.Device, errors.EdgeX) {
+	device = onvifClient.initCustomMetadata(device)
```

Comment:
if you make the `initCustomMetadata` take a pointer, change this (and all other usages) to 
```suggestion
	onvifClient.initCustomMetadata(&device)
```